internal/graph/check_test.go

```diff
@@ -581,3 +578,169 @@ condition condition1(param1: string) {
 	require.NoError(t, err)
 	require.False(t, resp.Allowed)
 }
+
+func TestUnionCheckFuncReducer(t *testing.T) {
+	ctx := context.Background()
+
+	concurrencyLimit := uint32(10)
+
+	falseHandler := func(context.Context) (*ResolveCheckResponse, error) {
+		return &ResolveCheckResponse{
+			Allowed:            false,
+			ResolutionMetadata: &ResolutionMetadata{},
+		}, nil
+	}
+
+	trueHandler := func(context.Context) (*ResolveCheckResponse, error) {
+		return &ResolveCheckResponse{
+			Allowed:            true,
+			ResolutionMetadata: &ResolutionMetadata{},
+		}, nil
+	}
+
+	t.Run("if_no handlers,_return_allowed:false", func(t *testing.T) {
+		resp, err := union(ctx, concurrencyLimit)
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("if_one_handler_is_truthy_and_others_are_falsy,_return_allowed:true", func(t *testing.T) {
+		resp, err := union(ctx, concurrencyLimit, trueHandler, falseHandler, falseHandler)
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	})
+
+	t.Run("if_all_handlers_are_falsy,_return_allowed:false", func(t *testing.T) {
+		resp, err := union(ctx, concurrencyLimit, falseHandler, falseHandler, falseHandler)
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("if_a_handler_errors_but_other_handler_are_truthy,_return_allowed:true", func(t *testing.T) {
+		depthExceededHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return nil, ErrResolutionDepthExceeded
+		}
+
+		resp, err := union(ctx, concurrencyLimit, depthExceededHandler, trueHandler)
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	})
+
+	t.Run("if_a_handler_errors_but_other_handler_is_falsy,_return_error_and_allowed:false", func(t *testing.T) {
+		depthExceededHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return nil, ErrResolutionDepthExceeded
+		}
+
+		resp, err := union(ctx, concurrencyLimit, depthExceededHandler, falseHandler)
+		require.ErrorIs(t, ErrResolutionDepthExceeded, err)
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("if_a_handler_errors_with_cycle_but_other_handler_is_falsy,_return_allowed:false_with_a_nil_error", func(t *testing.T) {
```

Comment:
this test and the one below it are the only one that have handlers throwing `ErrCycleDetected`. but they both have _two_ handlers.

can you add a test for where there is only _one_ handler, and it throws `ErrCycleDetected`? in that case, following what @rhamzeh said here: https://github.com/openfga/openfga/pull/1372#pullrequestreview-1885724198

we should return the error, no?

Reply:
never mind - if one handler returns ErrCycleDetected the final outcome is falsy

same if you have N handlers, if they all return ErrCycleDetected,  the final outcome is falsy